torch/csrc/distributed/c10d/UCCUtils.cpp

```diff
@@ -0,0 +1,274 @@
+  worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+  worker_params.thread_mode = UCS_THREAD_MODE_MULTI;
+  st = ucp_worker_create(context, &worker_params, &worker);
+  if (st != UCS_OK) {
+    TORCH_UCC_LOG_ERROR(
+        TORCH_UCC_INIT,
+        c10::str("UCX failed to create UCP worker:", ucs_status_string(st)));
+    ucp_cleanup(context);
+    throw std::runtime_error(ucs_status_string(st));
+  }
+}
+
+void CommUCX::progress() {
+  ucp_worker_progress(worker);
+}
+
+void CommUCX::free_request(ucc_coll_req_h request) {
+  request->status = UCC_INPROGRESS;
+  ucp_request_free(request);
+}
+
+CommUCX::~CommUCX() {
+  if (worker != nullptr) {
+    ucp_worker_destroy(worker);
+  }
+  if (context != nullptr) {
+    ucp_cleanup(context);
+  }
+  worker = nullptr;
+  context = nullptr;
+}
+
+ucc_status_t oob_allgather(
+    void* sbuf,
+    void* rbuf,
+    size_t msglen,
+    void* coll_info,
+    void** req) {
+  torch_ucc_oob_coll_info_t* info =
+      reinterpret_cast<torch_ucc_oob_coll_info_t*>(coll_info);
+  std::vector<uint8_t> val = std::vector<uint8_t>(
+      reinterpret_cast<uint8_t*>(sbuf),
+      reinterpret_cast<uint8_t*>(sbuf) + msglen);
+  try {
+    info->store->set(info->getKey("teamr" + std::to_string(info->rank)), val);
+    info->rbuf = rbuf;
+    info->msglen = msglen;
+    *req = coll_info;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "(oob_allgather) Caught exception in Store Operation .. "
+               << "[" << ex.what() << "]";
+    return UCC_ERR_NO_MESSAGE;
+  }
+  return UCC_OK;
+}
+
+ucc_status_t oob_allgather_test(void* req) {
+  torch_ucc_oob_coll_info_t* info =
+      reinterpret_cast<torch_ucc_oob_coll_info_t*>(req);
+
+  try {
+    for (int r = 0; r < info->size; r++) {
+      if (!info->store->check({info->getKey("teamr" + std::to_string(r))})) {
+        return UCC_INPROGRESS;
+      }
+    }
+    for (int r = 0; r < info->size; r++) {
+      std::vector<uint8_t> data =
+          info->store->get(info->getKey("teamr" + std::to_string(r)));
+      memcpy(
+          (void*)((ptrdiff_t)info->rbuf + info->msglen * r),
+          data.data(),
+          info->msglen);
+    }
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "(oob_allgather) Caught exception in Store Operation .. "
+               << "[" << ex.what() << "]";
+    return UCC_ERR_NO_MESSAGE;
+  }
+  return UCC_OK;
+}
+
+ucc_status_t oob_allgather_free(void* req) {
+  torch_ucc_oob_coll_info_t* info =
+      reinterpret_cast<torch_ucc_oob_coll_info_t*>(req);
+  try {
+    int num_done = info->store->add({info->getKey("ag_done")}, 1);
+    if (num_done == info->size) {
+      info->store->deleteKey(info->getKey("ag_done"));
+      for (int r = 0; r < info->size; r++) {
+        info->store->deleteKey(info->getKey("teamr" + std::to_string(r)));
+      }
+      for (int r = 0; r < info->size; r++) {
+        info->store->add({info->getKey("ag_free" + std::to_string(r))}, 1);
+      }
```

Comment:
Is there a reason not to do it in the same loop? Also, consider using `c10::irange` primitive
```suggestion
      for (const auto r: c10::irange(info->size)) {
        info->store->deleteKey(info->getKey("teamr" + std::to_string(r)));
        info->store->add({info->getKey("ag_free" + std::to_string(r))}, 1);
      }
```

Reply:
I added a comment that it is required as separate loop.